internal/worker/sshserver/session.go

```diff
@@ -6,13 +6,113 @@ package sshserver
 import (
 	"github.com/gliderlabs/ssh"
 
+	"github.com/juju/errors"
 	"github.com/juju/juju/core/virtualhostname"
+	gossh "golang.org/x/crypto/ssh"
+
+	"github.com/juju/juju/state"
 )
 
 type stubSessionHandler struct{}
 
 // Handle is a stub implementation of the SessionHandler interface.
-// It currently does nothing but will be used to proxy a user's SSH session
-// to a target machine.
+// It currently does nothing but will be replaced with a real implementation
+// that proxies user's requests to the target unit or machine.
 func (s *stubSessionHandler) Handle(session ssh.Session, destination virtualhostname.Info) {
 }
+
+// SSHConnecter is an interface that defines the methods required to
+// connect to a remote SSH server and determine the model type of the target.
+type SSHConnecter interface {
+	Connect(destination virtualhostname.Info) (*gossh.Client, error)
+}
+
+type sessionHandler struct {
+	connector SSHConnecter
+	modelType state.ModelType
+	logger    Logger
+}
+
+// Handle proxies a user's SSH session to a target unit or machines.
+// Connections to machine will be proxied to the machine's SSH server.
+// Connections to k8s units will be proxied through the k8s API server.
+func (s *sessionHandler) Handle(session ssh.Session, destination virtualhostname.Info) {
+	handleError := func(err error) {
+		s.logger.Errorf("proxy failure: %v", err)
+		_, _ = session.Stderr().Write([]byte(err.Error() + "\n"))
+		_ = session.Exit(1)
+	}
+
+	switch s.modelType {
+	case state.ModelTypeCAAS:
+		if err := s.K8sSessionProxy(session); err != nil {
+			err = errors.Annotate(err, "failed to proxy k8s session")
+			handleError(err)
+		}
+	case state.ModelTypeIAAS:
+		if err := s.MachineSessionProxy(session, destination); err != nil {
+			err = errors.Annotate(err, "failed to proxy machine session")
+			handleError(err)
+		}
+	default:
+		handleError(errors.Errorf("unknown model type %s", s.modelType))
+	}
+}
+
+func (s *sessionHandler) K8sSessionProxy(sess ssh.Session) error {
```

Comment:
does this method need to be exported? if yes, then godoc, please

Reply:
Changed to no longer be exported.